services/venice-controller/src/main/java/com/linkedin/venice/controller/kafka/protocol/serializer/SchemaDiffTraverser.java

```diff
@@ -0,0 +1,421 @@
+    };
+  }
+
+  private boolean handleMissingTargetField(
+      Object object,
+      Schema.Field currentField,
+      AtomicReference<String> errorMessage) {
+    if (isNestedType(currentField.schema())) {
+      Object fieldDefaultValue = AvroSchemaUtils.getFieldDefault(currentField);
+      if (fieldDefaultValue == null) {
+        errorMessage.set(
+            "Field: " + formatFieldName(currentField)
+                + " is not in the target schema, and object is non-null. Actual value: " + object);
+        return true;
+      }
+      if (!fieldDefaultValue.equals(object)) {
+        errorMessage.set(
+            "Field: " + formatFieldName(currentField) + " contains non-default value. Actual value: " + object
+                + ". Default value: " + fieldDefaultValue + " or " + null);
+        return true;
+      }
+      return false;
+    }
+
+    return isNonDefaultValue(object, currentField, errorMessage);
+  }
+
+  private boolean hasSchemaTypeMismatch(
+      Schema.Field currentField,
+      Schema.Field targetField,
+      AtomicReference<String> errorMessage) {
+    if (currentField.schema().getType() != targetField.schema().getType()) {
+      errorMessage.set("Field: " + formatFieldName(currentField) + " has different types");
+      return true;
+    }
+    return false;
+  }
+
+  private boolean checkSchemaSpecificCases(
+      Object object,
+      Schema.Field currentField,
+      Schema.Field targetField,
+      AtomicReference<String> errorMessage) {
+    switch (currentField.schema().getType()) {
+      case ENUM:
+        return isNewEnumValue(object, currentField, targetField, errorMessage);
+      case FIXED:
+        if (currentField.schema().getFixedSize() != targetField.schema().getFixedSize()) {
+          errorMessage.set("Field: " + formatFieldName(currentField) + " has different fixed size");
+          return true;
+        }
+        return false;
+      default:
+        return isNonDefaultValue(object, currentField, errorMessage);
+    }
+  }
+
+  private String formatFieldName(Schema.Field field) {
+    return field.name().replace("_", ".");
+  }
+
+  private boolean isNewEnumValue(
+      Object object,
+      Schema.Field currentField,
+      Schema.Field targetField,
+      AtomicReference<String> errorMessage) {
+    List<String> currentSymbols = currentField.schema().getEnumSymbols();
+    List<String> targetSymbols = targetField.schema().getEnumSymbols();
+    if (targetSymbols.stream().anyMatch(symbol -> !currentSymbols.contains(symbol) && symbol.equals(object))) {
+      errorMessage.set("Field: " + currentField.name().replace("_", ".") + " contains new enum value: " + object);
+      return true;
+    }
+
+    return false;
+  }
+
+  private boolean isNonDefaultValue(Object value, Schema.Field field, AtomicReference<String> errorMessage) {
+    if (value == null)
+      return false;
+
+    Map<Schema.Type, Object> defaultValues = new HashMap<>();
```

Comment:
actually instead of allocating a new map just to check for defaults, can you use switch statement?